Refactors `src/package.rs`

xfbs opened the pull request:
This PR refactors the module `package` in the buffrs CLI. Since I am looking to integrate the `validation` crate with `buffrs`, I saw some low-hanging fruits for code improvement. On a high level, the improvements introduced by this PR address the following:

- *async correctness*: not using blocking code in the path of async code. This is quite easy to get wrong in Rust, as of yet there is not much compiler help to avoid this (to my knowledge). In certain cases, it can be okay to mix the two (using a blocking mutex in a hot loop, writing a CLI that is single-threaded). But to build robust, reusable code we should try to get this right.
- *error handling*: error handling is something that can always use some love. This PR adds some error types in strategic places, mostly to make testing easier.
- *test coverage*: this PR attempts to increase net code coverage to 80%.
- *encapsulation*: trying to not leak implementation details when it is not necessary to leak them. This allows us freedom to change the implementation without breaking too many things.
- *robustness*: trying to avoid hidden assumptions by making them explicit. One example for this is the reliance on the current working directory in some of the code paths. 

I try to keep the changes in here to be mostly non-opinionated, so that they are uncontroversial. 

**Edit: this PR is now ready**. Here is a list of the major changes I have made here:

- [Consolidated the `impl PackageStore` blocks](https://github.com/helsing-ai/buffrs/pull/132/commits/3f387ead357b6a57a9bc7d0e09bc87618498aafd). There were multiple of them, I just squashed them together into a single one.
- [Removes superfluous `TryFrom` implementations](https://github.com/helsing-ai/buffrs/pull/132/commits/3e9a3b94ee736934ac2938352308a1aa9bdab624). `TryFrom` and `FromStr` are handy, but we should only implement the ones we need. For cases where a string slice suffices, `FromStr` is the way to go. For cases where ownership is needed, `TryFrom<String>` is the way to go. Always prefer `TryFrom<&str>` over `TryFrom<&String>`, because the former is more generic (the latter is just one specific container type for a `&str`), but usually neither are useful because what you really want is `FromStr`. Here I just removed some implementations that we are not using and that do not add value.
- [Implements custom error type for `PackageName` parsing](https://github.com/helsing-ai/buffrs/pull/132/commits/4e0e377efdefbf87146a8ce728bb66d48a18834e). Here I implemented a custom error type, `PackageNameError`, using `thiserror`. This error type helps write useful unit tests, `miette` is neat for displaying but not so much for "low-level" errors if that makes sense.
- [Changes `Deref` of `PackageName` to target `str`](https://github.com/helsing-ai/buffrs/pull/132/commits/8bfe6e7fc075e6c67eba6859231103ae9c8e0d4d). In general, similar as with the `TryFrom` implementation, always prefer `&str` over `&String`. The reason is that `&str` is a generic string (can be backed by any container). Here, I just swapped out the `Target` of the `Deref` and made it work.
- [Moves resolving code into `src/resolver.rs`](https://github.com/helsing-ai/buffrs/pull/132/commits/9c4c4c7f482e29cb4b2536e05126194fd1091ccc): in having *one module do one thing*, I felt that resolving was different enough to warrant putting it into it's own module.
- [Derive `Default` for `PackageType`](https://github.com/helsing-ai/buffrs/pull/132/commits/88b3c9e17c3b7433781380ec96ff8593affd4d0b). This type already had a manually-implemented `Default`, but I always prefer `derive` implementations over hand-rolled ones unless absolutely necessary, for maintainability and brevity reasons.
- [Don't use `unwrap_or_default`](https://github.com/helsing-ai/buffrs/pull/132/commits/e54e764410275b1197ba514732dc8c6c30f3935c): Rust has amazing error handling, but there are cases where things actually *should* panic. This is one of those, so I have switched to using `unreachable` rather than silently converting to an empty string. If this ever breaks, we certainly want to know about it!
- [Removes `From<&Package>` implementation](https://github.com/helsing-ai/buffrs/pull/132/commits/39f586bed491ef014a5b11bebd4cab0428cc73d3): this implementation was not used and does not add any value, because the data is available in a public field.
- [Add unit tests for `PackageType`](https://github.com/helsing-ai/buffrs/pull/132/commits/1a56a1f2f17c6cd73b935e65a45becf0dbbd8c48): this cleans up `PackageType` a bit, adding unit tests, making the code a bit clearer.
- [Remove implicit assumptions from `PackageStore`](https://github.com/helsing-ai/buffrs/pull/132/commits/af86bcab5131b8fa93f14640901e75cb5dea4dbf): The `PackageStore` currently heavily relies on the current working directory. I do not like using mutable global state, and also for testability purposes this is not great. So instead I added a field that contains the root path that all relative paths are resolved using.
- [Don't read `Manifest` in `PackageStore::release`](https://github.com/helsing-ai/buffrs/pull/132/commits/2dd5b8b83f74f917783ba0f2d839f2ebf078e723): in the interest of decoupling things, we don't really want the `PackageStore` to have to interact with manifests (such as reading them). Rather, we want to decouple them by having it such that the `Manifest` is a method argument that we simply pass through.
- [Decouples `PackageStore` from compression](https://github.com/helsing-ai/buffrs/pull/132/commits/892256735680da4796e1477cb204880548ac2782): similar as with the `Manifest`, our `PackageStore` should only ever deal with package storage things and not with compression. So I have moved the compression-related code out into `Package` instead, this decouples things.

Before:

<img width="863" alt="test-coverage-before" src="https://github.com/helsing-ai/buffrs/assets/786420/af1313ed-83f5-4a7d-a493-124c61fc0dcd">

After:

<img width="863" alt="test-coverage-after" src="https://github.com/helsing-ai/buffrs/assets/786420/93523b1b-99d1-4d05-81bf-a75d3bc0aed7">

I did not quite get the test coverage up to 80%, but we did hit about 77% (with some cheating because I have moved out the resolver code). 

This PR turned out rather wordy, but I hope it all makes sense! I think in a next iteration I could tackle the testing situation a bit more.

mara-schulke commented on src/package.rs:
```diff
@@ -13,78 +13,96 @@
-pub struct PackageStore;
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PackageStore {
+    root: PathBuf,
```
We had a similar PR by @uschi2000

See the comments of #24 

I have to say, i still dislike the oop complexity associated with this, i get that the imperative approach is not maintainable longterm – can we find a functional architecture?

xfbs replied:
How do you define *OOP complexity*?

Can you give an example of how a functional architecture would look like here that would model this better?

mara-schulke replied:
`struct` / _instance_ orientation rather than dataflow orientation

the former results in (potentially huge) interlinked objects, flying around with limited composability where as the latter typically separated data transformation (logic) from io and results in better composability, testability and decomp at the same time

Imo (these functions) where all, mostly unrelated from each other, as they are IO heavy and do very little data transformation. A functional design could be to separate the IO part completely

So in this case, as most of the functions do only IO, i dont see the need for them to "belong" to a struct, rather a module, and remove the functions that nest their IO behavior.

---

As i said on rob's pr:

We also dont really have a need currently for making the path configurable, if thats the case the change is valid, but until then it is unnecessary complexity (imo).

mara-schulke commented:
Overall i like the split of "resolver" vs "package" – i would strive for as functional / pure designs as possible (i know that is not the case right now due to pragmatism) – but if we refactor stuff, i would want to move away from oop / imperative code as much as sensibly possible

mara-schulke reviewed (changes requested):
Overall i like the split of "resolver" vs "package" – i would strive for as functional / pure designs as possible (i know that is not the case right now due to pragmatism) – but if we refactor stuff, i would want to move away from oop / imperative code as much as sensibly possible

mara-schulke commented on src/package.rs:
```diff
@@ -13,78 +13,96 @@
                 .await
                 .into_diagnostic()
-                .wrap_err(miette!("failed to create {dir} directory"))
+                .wrap_err(miette!("failed to create {dir:?} directory"))
```
```suggestion
                .wrap_err(miette!("failed to create {} directory", dir.display()))
```

xfbs commented:
Discussed this today, the resolution here is that I will add a context to the `main.rs` which contains an instance (perhaps an `Arc<PackageStore>`) to avoid instances of this being created all over the place. Most of the code should have access to this context struct and it can be used to add other shared, immutable state.

Besides that, perhaps in a future PR, we should try to make sure that we don't mix pure and non-pure code all over the place. To some extend, this PR already addresses that, but there is still opportunities to improve that in future PRs.

mara-schulke commented on src/command.rs:
```diff
@@ -70,9 +72,10 @@ pub async fn init(kind: PackageType, name: Option<PackageName>) -> miette::Resul
     manifest.write().await?;
 
-    PackageStore::create()
+    PackageStore::create(std::env::current_dir().unwrap_or_else(|_| ".".into()))
```
Can we change the signature so people dont have to do this?

like `PackageStore::current().try_create()` or something like this?

xfbs replied:
I thought we wanted to aim to be functional?

mara-schulke commented on src/command.rs:
```diff
@@ -132,14 +136,22 @@ pub async fn remove(package: PackageName) -> miette::Result<()> {
 
-    PackageStore::uninstall(&dependency.package).await.ok(); // temporary due to broken test
+    PackageStore::current()
+        .into_diagnostic()?
```
Why is this needed? `PackageStore::xyz` should all return a `miette::Result` which you can just use with `?` directly

xfbs replied:
This code is not from me, it is from @asmello, he said this was a workaround for a broken unit test.

xfbs replied:
Oh wait, I misread. 

No, PackageStore returns a fine-grained error, which in this case is an `std::io::Error`. You cannot use the `?` operator with `miette::Result`. You always need to convert any errors into miette errors work `into_diagnostic()`.

mara-schulke commented on src/command.rs:
```diff
@@ -188,7 +200,11 @@ pub async fn publish(
-    let package = PackageStore::release().await?;
+    let manifest = Manifest::read().await?;
+    let package = PackageStore::current()
+        .into_diagnostic()?
```
Same here

xfbs replied:
Same reason: `current()` only checks some stuff that may return `std::io::Error`. We always try to return the most specific error type if we can.

`miette::Result` does not work like `eyre::Result` or `anyhow::Result`. You cannot just cast any error into a `miette::Error` with `?`, you have to explicitly convert it using `into_diagnostic()`.

We don't want to change the return error of `PackageStore::current()`, because then we would make the errors less useful and we would make testing harder than it needs to be. In my understanding, `miette::Result` is mostly useful for pretty-printing errors.

So we will have to deal with these `into_diagnostic` calls in certain places, as far as I can tell.

mara-schulke commented on src/command.rs:
```diff
@@ -188,7 +200,11 @@ pub async fn publish(
+    let manifest = Manifest::read().await?;
+    let package = PackageStore::current()
+        .into_diagnostic()?
+        .release(manifest)
```
Can we split the releasing out? As we figured out yesterday its not really related to the package store

mara-schulke replied:
Maybe we can put it into `Package` for now` Like `Package::release` or alike

xfbs replied:
Sure, but that is something that should go into a separate PR. I don't want this one to be too big.

mara-schulke commented on src/generator.rs:
```diff
@@ -126,33 +123,32 @@ impl Generator {
                 ":: compiled {} [{}]",
                 manifest.package.name,
-                location.display()
+                store.proto_path().display()
```
Can we rename the methods to something snappier? `store.dir()` and `store.vendor_dir()`? 

I still think that the "proto_path" is not related to the store, so:

proto/* -> the current package
proto/vendor/* -> the "package store"

So if we would split out the logic for accessing the first path that would be great. I think the weirdness that the vendor dir lives inside of the proto dir creates these blurry borders

xfbs replied:
I stuck with the names that were used before: `PROTO_DIR` and `PROTO_VENDOR_DIR`. I'll create an issue for this.

mara-schulke commented on src/package.rs:
```diff
@@ -13,78 +13,96 @@
+    }
+
+    /// Path to the `proto` directory.
+    pub fn proto_path(&self) -> PathBuf {
```
Imo this should not return the pathbuf? can we return a borrowed path to prevent an alloc?

xfbs replied:
It returns the `root` path with a directory appended.

I would not really worry about allocations here, the cost is negligible and changing it would be a premature optimization. But we could add an issue for it for someone to tackle at some point.

Perhaps the package store itself could just contain the path to the vendor dir.

mara-schulke replied:
Its more about semantics than allocations, returning a borrowed path makes the lifetime of the path the same, or shorter, than the store

But also i dont really care

mara-schulke commented on src/package.rs:
```diff
@@ -173,16 +184,84 @@ impl PackageStore {
+}
+
+#[test]
+fn can_get_proto_path() {
```
can you group all tests that you wrote in `#[cfg(test)] mod test` as usual?

xfbs replied:
Sure, I'll do that.

mara-schulke commented on src/package.rs:
```diff
@@ -262,88 +317,11 @@ impl PackageStore {
-
-    fn try_from(tgz: Bytes) -> Result<Self, Self::Error> {
+    /// Load a package from a precompressed archive.
+    pub fn load(tgz: Bytes) -> miette::Result<Self> {
```
Imo lets not do this, also we dont load, but rather parse

xfbs replied:
I don't quite follow?

mara-schulke replied:
So, you moved this from `TryFrom<Bytes>` into `::load` and i think there is no reason, as we dont reuse this, and the semantic meaning of load is weird – we dont do IO here

xfbs replied:
Well, having any kind of complex logic inside of a trait implementation is a strong code smell — so I had to move this into a method. I could make the method non-public and rename it. But I would definitely not leave logic inside the `impl TryFrom<Bytes>`, that is not a good idea.

xfbs replied:
I'll rename it to parse and make it non-pub perhaps.

mara-schulke commented on src/package.rs:
```diff
@@ -385,16 +363,44 @@ impl TryFrom<Bytes> for Package {
+    }
 }
 
-impl From<&Package> for Bytes {
```
Why has this been removed?

xfbs replied:
It is a red flag to have a `From` for a field unless the field semantically represents that type. A package is a package, not something that is equal to some bytes. Also, most importantly, it was not used. Better to use `package.tgz` explicitly.

mara-schulke replied:
A package is strictly equal to bytes

mara-schulke replied:
So, you can do `Bytes -> Package -> Bytes -> Package -> Bytes` all day

The manifest is only parsed out in the `TryFrom` to prevent the need of ungzipping on every manifest access

mara-schulke commented on src/package.rs:
```diff
@@ -429,73 +436,139 @@ impl FromStr for PackageType {
+    }
+
+    /// New package name from string.
+    pub fn new<S: Into<String>>(value: S) -> Result<Self, PackageNameError> {
```
Can this go to the top of the impl block?

mara-schulke replied:
And return a miette::Result

mara-schulke commented on src/package.rs:
```diff
@@ -429,73 +436,139 @@ impl FromStr for PackageType {
-    type Error = miette::Report;
+/// Errors that can be generated parsing [`PackageName`][], see [`PackageName::new()`][].
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum PackageNameError {
```
```suggestion
enum PackageNameError {
```

mara-schulke replied:
And please, if possible, derive miette diagnostics

xfbs replied:
I strongly disagree with making this private.

mara-schulke commented on src/package.rs:
```diff
@@ -429,73 +436,139 @@ impl FromStr for PackageType {
 
 impl FromStr for PackageName {
-    type Err = miette::Report;
+    type Err = PackageNameError;
```
```suggestion
    type Err = miette::Report;
```

xfbs replied:
I *heavily* disagree with using `miette::Report` here. `miette` is an error *report* library. It helps you create reports for errors. Meaing: you can use it to convert `low-level` errors into something that is pretty-printable. This is not the place to use it, that is why this PR has removed it. 

The idea is: you use fine-grained errors in your application, and at some point before presenting it to the user, you can turn your error into a report for presentation.

At this point, all we care about is parsing `PackageName`s and returning a useful error type that tells you *exactly why* the parsing failed. Which means an `enum` with fields and an `std::error::Error` implementation.
Downstream, you are then free to turn this into an error *report* using `miette`. But if we have this useful error type, we should use it instead. A `FromStr` should never return a `miette::Report`, that is a big code smell (to me).

mara-schulke replied:
Please see #95 and #103

mara-schulke replied:
This implementation is public, meaning that you leak this to the outside world

mara-schulke reviewed (changes requested):
So overall, yes we can merge this, but i found a few nits regarding style, and a few issues in terms of unnecessarily moved code, leaking of impl details (see the error related comments, maybe take a look at #95 and #103 for ref) and ergonomic issues (especially the into_diagnostic thing)

xfbs commented on src/package.rs:
```diff
@@ -13,78 +13,96 @@
+    }
+
+    /// Open given directory.
+    pub fn open(path: &Path) -> Self {
```
Have created an issue for this.

mara-schulke replied:
I think this should be resolved within this PR if im honest 😅  But okay

xfbs commented on src/command.rs:
```diff
@@ -223,12 +239,11 @@ pub async fn install(credentials: Credentials) -> miette::Result<()> {
-                &resolved.package.name()
-            ))?;
+        let store = PackageStore::current().into_diagnostic()?;
+        store.unpack(&resolved.package).await.wrap_err(miette!(
```
I don't quite follow?

mara-schulke replied:
```suggestion
    PackageStore::current().into_diagnostic()?.unpack(&resolved.package).await.wrap_err(miette!(
```

mara-schulke replied:
(Formatting is broken)

xfbs commented:
I've addressed the feedback:
- I've created some issues for some of the issues, because I do not want this PR to get any larger than it already is. Prefer to work in small chunks. 
- Some of the issues I have addressed
- Some of the issues I disagree with, and I have left a comment explaining the rationale.
